src/domain/usecases/CopyProgramStageDataValuesUseCase.ts

```diff
@@ -0,0 +1,175 @@
+        });
+
+        const errors = dataElementPairs.filter(pair => typeof pair === "string");
+        if (!_.isEmpty(errors)) throw new Error(errors.join("\n"));
+
+        return dataElementPairs.filter((pair): pair is DataElementPair => typeof pair !== "string");
+    }
+
+    private saveReport(
+        path: string,
+        dataElementPairs: DataElementPair[],
+        programStageId: string,
+        eventsWithNewDataValues: ProgramEvent[]
+    ) {
+        const dataElementLines = dataElementPairs.map(
+            ([source, target]) =>
+                `Source DataElement: ${source.id} (${source.name}), Target DataElement: ${target.id} (${target.name})`
+        );
+
+        const eventLines = eventsWithNewDataValues.map(event => {
+            const dataValueLines = dataElementPairs.flatMap(([source, target]) => {
+                const sourceValue = event.dataValues.find(dv => dv.dataElement.id === source.id)?.value;
+                const status = sourceValue ? `(${sourceValue})` : undefined;
+                return status ? [`\tCopy ${source.id} to ${target.id} ${status}`] : [];
+            });
+
+            return `Event ID: ${event.id}, OrgUnit ID: ${event.orgUnit.id}\n${dataValueLines.join("\n")}`;
+        });
+
+        const content = [
+            "Program Stage ID: " + programStageId,
+            dataElementLines.join("\n"),
+            "Number of events: " + eventsWithNewDataValues.length,
+            eventLines.join("\n"),
+        ].join("\n\n");
+
+        fs.writeFileSync(path, content);
+        log.info(`Written report: ${path}`);
+    }
+}
+
+function checkDataElementTypes(dePairs: DataElementPair[]) {
+    const typeMismatchErrors = dePairs
+        .filter(([source, target]) => source.valueType !== target.valueType)
+        .map(([source, target]) => `Data elements [${source.id}, ${target.id}] do not have the same type.`);
+
+    if (!_.isEmpty(typeMismatchErrors)) throw new Error(typeMismatchErrors.join("\n"));
+}
+
+function checkTargetDataValuesAreEmpty(events: ProgramEvent[], targetIds: Id[]) {
+    const eventsWithNonEmptyTargetDataValues = _(events)
+        .map(event => {
+            const nonEmpty = event.dataValues
+                .filter(dv => targetIds.includes(dv.dataElement.id))
+                .filter(dv => Boolean(dv.value))
+                .map(dv => `\tTarget DataElement: ${dv.dataElement.id}, Value: ${JSON.stringify(dv.value)}`)
+                .join("\n");
+
+            return _.isEmpty(nonEmpty) ? undefined : `Event ID: ${event.id}, Values: \n${nonEmpty}`;
+        })
+        .compact()
+        .join("\n");
+
+    const error = `Some data values of the destination data elements are not empty:\n${eventsWithNonEmptyTargetDataValues}`;
+    if (eventsWithNonEmptyTargetDataValues) throw new Error(error);
+}
+
+export type CopyProgramStageDataValuesOptions = {
+    programStageId: string;
+    dataElementIdPairs: [Id, Id][]; // [sourceDataElementId, targetDataElementId]
```

Comment:
Working with tuples should be avoided, all the code interacting with them will be uglier. And you see, you had to add a comment about what it holds. So the most declarative structure is an object: `type DataElementMapping = {sourceId: Id, targetId: Id}`